crypto/fipsmodule/rand/entropy/tree_drbg_jitter_entropy.c

```diff
@@ -0,0 +1,465 @@
+  }
+
+  entropy_source->state = tree_jitter_drbg;
+
+  return 1;
+}
+
+#if defined(_MSC_VER)
+#pragma section(".CRT$XCU", read)
+static void tree_jitter_free_global_drbg(void);
+static void windows_install_tree_jitter_free_global_drbg(void) {
+  atexit(&tree_jitter_free_global_drbg);
+}
+__declspec(allocate(".CRT$XCU")) void(*tree_jitter_drbg_destructor)(void) =
+    windows_install_tree_jitter_free_global_drbg;
+#else
+static void tree_jitter_free_global_drbg(void) __attribute__ ((destructor));
+#endif
+
+// The memory life-time for thread-local seed DRBGs is handled differently
+// compared to the global seed DRBG (and Jitter Entropy instance). The frontend
+// DRBG thread-local destuctors will invoke |tree_jitter_free_thread_drbg| using
+// their reference to it. The global seed DRBG and Jitter Entropy instance will
+// be released by a destructor. This ensures that the global seed DRBG life-time
+// extends to the entire process life-time if the lazy initialization happened.
+// Obviously, any dlclose on AWS-LC will release the memory early but that's
+// correct behaviour.
+
+// tree_jitter_free_global_drbg frees the memory allocated for the global seed
+// DRBG and Jitter Entropy instance.
+static void tree_jitter_free_global_drbg(void) {
+
+  CRYPTO_STATIC_MUTEX_lock_write(global_seed_drbg_lock_bss_get());
+
+  struct tree_jitter_drbg_t *global_tree_jitter_drbg = *global_seed_drbg_bss_get();
+  if (*global_seed_drbg_bss_get() == NULL) {
+    return;
+  }
+
+  if (global_tree_jitter_drbg->is_global != 1) {
+    // Should not happen.
+    abort();
+  }
+
+  jent_entropy_collector_free(global_tree_jitter_drbg->jitter_ec);
+  OPENSSL_free(global_tree_jitter_drbg);
+
+  *global_seed_drbg_bss_get() = NULL;
+
+  CRYPTO_STATIC_MUTEX_unlock_write(global_seed_drbg_lock_bss_get());
+}
+
+// tree_jitter_zeroize_thread_drbg frees the thread-local seed DRBG
+// associated with the entropy source |entropy_source|.
+void tree_jitter_free_thread_drbg(struct entropy_source_t *entropy_source) {
+
+  GUARD_PTR_ABORT(entropy_source);
+
+  struct tree_jitter_drbg_t *tree_jitter_drbg =
+    (struct tree_jitter_drbg_t *) entropy_source->state;
+  if (tree_jitter_drbg == NULL) {
+    return;
+  }
+
+  OPENSSL_free(tree_jitter_drbg);
+  entropy_source->state = NULL;
+}
+
+// Per ISO/IEC 19790-2012 7.9.7 "zeroization" can be random data just not other
+// SSP/CSP's. The Jitter Entropy instance doesn't have any practical state; it's
+// a live entropy source. The zeroization strategy used for the DRBG's is to
+// reseed with random data, that in turn, will override all states in the tree
+// with random data. The zeroization of the tree DRBG executes after the
+// frontend DRBGs have been locked - they can't release any generated output.
+// Therefore, the randomness generation layer ensures that no output from an the
+// tree DRBG is used to generate any output that is later released. Randomizing
+// the tree DRBG states therefore effectively "zeroize" the state.
```

Comment:
If the frontends are locked why not just zeroize the tree state? What could be using these if the frontends are locked forever?

Reply:
We do not assume that the entropy source cannot be queried. It's only the usage of the DRBG frontend state that is assumed to be locked.

Reply:
Why not? It seems weird that we're treating the two differently. Why lock the front end DRBGs but randomize the tree? Why not vice versa, or why not treat them the same?

Reply:
This goes back to the thing about powerpc and system calls. See `RAND_bytes_core()`. If I did not have this condition, I could make this look prettier for sure.

But model-wise, it's only the object `CTR_DRBG_STATE` inside each thread-local `rand_thread_local_state` that is synchronised by the thread-local lock `state_clear_lock`.

I attempted to turn the entropy sources into an abstraction such that we could more easily configure them. Before they were intertwined throughout code-paths.

Reply:
Oh yeah, powerpc... But is that an issue here? During the shutdown once the DRBGs are locked we don't want to make any more syscalls right?

Reply:
It's not. But the locking for the thread-local DRBGs at the frontend layer has no impact on whether the thread-local entropy source can be invoked. So, I can't release memory. To not having to reason about it, I just leave it (but do "zeroise" it).